contrib/win32/win32compat/shell-host.c

```diff
@@ -196,47 +204,133 @@ ConSRWidth()
 	return consoleBufferInfo.srWindow.Right;
 }
 
+void
+InitKeyTranslateMap(void)
+{
+	int i;
+	for (i=0; i < ARRAYSIZE(keys); i++) {
+		keys[i].len = strlen(keys[i].incoming);
+	}
+}
+
+struct key_translation *
+FindKeyTransByMask(char prefix, const char * value, int vlen, char suffix)
+{
+	int i;
+	for (i=0; i < ARRAYSIZE(keys); i++) {
+		struct key_translation * k = &keys[i];
+		if (k->len < vlen + 2) continue;
+		if (k->incoming[0] != '\033') continue;
+		if (k->incoming[1] != prefix) continue;
+		if (k->incoming[vlen+2] != suffix) continue;
+		if (vlen <= 1 && value[0] == k->incoming[2])
+			return k;
+		if (vlen > 1 && strncmp(&k->incoming[2], value, vlen) == 0)
+			return k;
+	}
+	return NULL;
+}
+
+int
+GetVirtualKeyByMask(char prefix, const char * value, int vlen, char suffix)
+{
+	struct key_translation * pk;
+	pk = FindKeyTransByMask(prefix, value, vlen, suffix);
+	return pk ? pk->vk : 0;
+}
+
 /*
  * This function will handle the console keystrokes.
  */
-void 
-SendKeyStroke(HANDLE hInput, int keyStroke, char character) 
+void
+SendKeyStrokeEx(HANDLE hInput, int vKey, char character, DWORD ctrlState, BOOL keyDown)
 {
 	DWORD wr = 0;
 	INPUT_RECORD ir;
 
 	ir.EventType = KEY_EVENT;
-	ir.Event.KeyEvent.bKeyDown = TRUE;
-	ir.Event.KeyEvent.wRepeatCount = 1;
-	ir.Event.KeyEvent.wVirtualKeyCode = keyStroke;
-	ir.Event.KeyEvent.wVirtualScanCode = 0;
-	ir.Event.KeyEvent.dwControlKeyState = 0;
+	ir.Event.KeyEvent.bKeyDown = keyDown;
+	ir.Event.KeyEvent.wRepeatCount = 0;
+	ir.Event.KeyEvent.wVirtualKeyCode = vKey;
+	ir.Event.KeyEvent.wVirtualScanCode = MapVirtualKeyA(vKey, MAPVK_VK_TO_VSC);
+	ir.Event.KeyEvent.dwControlKeyState = ctrlState;
 	ir.Event.KeyEvent.uChar.UnicodeChar = 0;
 	ir.Event.KeyEvent.uChar.AsciiChar = character;
 
 	WriteConsoleInputA(hInput, &ir, 1, &wr);
+}
 
-	ir.Event.KeyEvent.bKeyDown = FALSE;
-	WriteConsoleInputA(hInput, &ir, 1, &wr);
+void 
+SendKeyStroke(HANDLE hInput, int keyStroke, char character)
+{
+	SendKeyStrokeEx(hInput, keyStroke, character, 0, TRUE);
+	SendKeyStrokeEx(hInput, keyStroke, character, 0, FALSE);
 }
 
 void 
-ProcessIncomingKeys(char * ansikey)
+ProcessIncomingKeys(char * ak)
 {
-	int keylen = strlen(ansikey);
+	int aklen = strlen(ak);
 
-	if (!keylen)
+	if (!aklen)
 		return;
 
+	/* Decode base VT100 commands */
 	for (int nKey=0; nKey < ARRAYSIZE(keys); nKey++) {
-		if (strcmp(ansikey, keys[nKey].incoming) == 0) {
+		if (keys[nKey].len == aklen && strcmp(ak, keys[nKey].incoming) == 0) {
 			SendKeyStroke(child_in, keys[nKey].vk, keys[nKey].outgoing);
 			return;
 		}
 	}
 
-	for (int i=0; i < keylen; i++)
-		SendKeyStroke(child_in, 0, ansikey[i]);
+	/* Decode keys when pressed ALT key */
+	if (aklen == 2 && ak[0] == '\033') {
```

Comment:
> Lets not add some code for which we didn't have the documentation...

About **ALT** key: http://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-Alt-and-Meta-Keys
Option **altSendsEscape** enable by default in Ubuntu 14.
Source code of linux **Xterm**:
- https://github.com/joejulian/xterm/blob/master/ptyx.h#L2093
- https://github.com/joejulian/xterm/blob/master/input.c#L1296-1299

> I couldn't find the MSDN documentation for the special keys when ctrl key is pressed with length 7

Tests on Ubuntu 14.04 OpenSSH-client:
- Example "ESC [ 1 ; **5** A": digit **5** signaled about CTRL pressed (CTRL + UpArrow). Documentation: http://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-PC-Style-Function-Keys (see latest table about **Modifiers**).
- Example "ESC [ [ 1 ; **3** A": digit **3** signaled about ALT pressed (ALT + UpArrow).

OpenSSH source code: https://github.com/PowerShell/openssh-portable/blob/latestw_all/contrib/win32/win32compat/tncon.h#L112-123